### PR TITLE
[Az.Resources] Fix unexplicable error message when subscription and scope are both null.

### PR DESCRIPTION
**Pull request by NoriZC**
## Description

Related issue: https://github.com/Azure/azure-powershell/issues/22716

In command `New-AzRoleDefinition`, if subscription is not in azcontext and scope is not provided, it will give 
> Get-AzRoleDefinition : Object reference not set to an instance of an object.

This error message is not explicable for user to understand. After updating, the error message will be 

> No subscription was found in the default profile and no scope was specified. Either specify a scope or use a tenant with a subscription to run the command.

Same issue should occur in commands below because of same code structure:
- `Get-AzRoleDefinition`
- `Remove-AzRoleDefinition`
- `Get-AzRoleAssignment`
- `New-AzRoleAssignment`
- `Remove-AzRoleAssignment`

So fixed for these commands as well.

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/Resources/Resources/RoleAssignments/RemoveAzureRoleAssignmentCommand.cs [hunk 1]
```diff
@@ -188,14 +188,19 @@ public override void ExecuteCmdlet()
                     ResourceGroupName = ResourceGroupName,
                     ResourceName = ResourceName,
                     ResourceType = ResourceType,
-                    Subscription = DefaultProfile.DefaultContext.Subscription.Id
+                    Subscription = DefaultProfile.DefaultContext.Subscription != null ? DefaultProfile.DefaultContext.Subscription.Id: null
```
Comment (BethanyZhou) [inline src/Resources/Resources/RoleAssignments/RemoveAzureRoleAssignmentCommand.cs:191]:
```suggestion
                    Subscription =DefaultProfile.DefaultContext.Subscription?.Id
```

File: src/Resources/Resources/RoleDefinitions/GetAzureRoleDefinitionCommand.cs [hunk 2]
```diff
@@ -61,15 +61,15 @@ public override void ExecuteCmdlet()
                 Scope = Scope,
                 ResourceIdentifier = new ResourceIdentifier
                 {
-                    Subscription = DefaultProfile.DefaultContext.Subscription.Id?.ToString()
+                    Subscription = DefaultProfile.DefaultContext.Subscription != null ? DefaultProfile.DefaultContext.Subscription.Id.ToString(): null
```
Comment (BethanyZhou) [inline src/Resources/Resources/RoleDefinitions/GetAzureRoleDefinitionCommand.cs:64]:
```suggestion
                    Subscription = DefaultProfile.DefaultContext.Subscription?.Id?.ToString()
```

File: src/Resources/Resources/RoleDefinitions/RemoveAzureRoleDefinitionCommand.cs [hunk 3]
```diff
@@ -89,10 +89,15 @@ public override void ExecuteCmdlet()
                 Scope = Scope,
                 ResourceIdentifier = new ResourceIdentifier
                 {
-                    Subscription = DefaultProfile.DefaultContext.Subscription.Id.ToString()
+                    Subscription = DefaultProfile.DefaultContext.Subscription != null ? DefaultProfile.DefaultContext.Subscription.Id.ToString() : null
```
Comment (BethanyZhou) [inline src/Resources/Resources/RoleDefinitions/RemoveAzureRoleDefinitionCommand.cs:92]:
```suggestion
                    Subscription = DefaultProfile.DefaultContext.Subscription?.Id?.ToString()
```

File: src/Resources/Resources/RoleAssignments/GetAzureRoleAssignmentCommand.cs [hunk 4]
```diff
@@ -245,6 +245,11 @@ public override void ExecuteCmdlet()
                 IncludeClassicAdministrators = IncludeClassicAdministrators.IsPresent,
             };
 
+            if (options.Scope == null && options.ResourceIdentifier.Subscription == null)
+            {
+                WriteTerminatingError(OdataHelper.ScopeAndSubscriptionBothNull(options.Scope, options.ResourceIdentifier.Subscription));
```
Comment (BethanyZhou) [inline src/Resources/Resources/RoleAssignments/GetAzureRoleAssignmentCommand.cs:250]: Why we judge (options.Scope == null && options.ResourceIdentifier.Subscription == null twice in L248 and OdataHelper.ScopeAndSubscriptionBothNull.

File: src/Resources/Resources/Helper/OdataHelper.cs [hunk 5]
```diff
@@ -47,6 +47,15 @@ public static bool IsAuthorizationDeniedException(this Common.MSGraph.Version1_0
             return false;
         }
 
+        public static string ScopeAndSubscriptionBothNull(string Scope, string Subscription)
+        {
```
Comment (BethanyZhou) [inline src/Resources/Resources/Helper/OdataHelper.cs:51]: This function should be not putted in OdataHelper.cs

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/Resources/Resources/Helper/RoleHelper.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
```
Comment (isra-fel) [inline src/Resources/Resources/Helper/RoleHelper.cs:1]: Please add license header for new files. You can copy from other .cs files

Comment (isra-fel) [reply src/Resources/Resources/Helper/RoleHelper.cs:1]: In fact, why not put it into a resx file (as a resource string?). Feel free to let me know if you don't know how

**Review by isra-fel — APPROVED** [review]

**Review by BethanyZhou — APPROVED** [review]

**Pull request merged**